internal/cloud/backend_taskStages.go

```diff
@@ -30,3 +52,142 @@ func (b *Cloud) runTaskStages(ctx context.Context, client *tfe.Client, runId str
 }
+
+func (b *Cloud) getTaskStageWithAllOptions(ctx *IntegrationContext, stageID string) (*tfe.TaskStage, error) {
+	options := tfe.TaskStageReadOptions{
+		Include: []tfe.TaskStageIncludeOpt{tfe.TaskStageTaskResults, tfe.PolicyEvaluationsTaskResults},
+	}
+	stage, err := b.client.TaskStages.Read(ctx.StopContext, stageID, &options)
+	if err != nil {
+		return nil, generalError("Failed to retrieve task stage", err)
+	} else {
+		return stage, nil
+	}
+}
+
+func (b *Cloud) runTaskStage(ctx *IntegrationContext, output IntegrationOutputWriter, stageID string) error {
+	var errs multiErrors
+
+	// Create our summarizers
+	summarizers := make([]taskStageSummarizer, 0)
+	ts, err := b.getTaskStageWithAllOptions(ctx, stageID)
+	if err != nil {
+		return err
+	}
+
+	if s := newTaskResultSummarizer(b, ts); s != nil {
+		summarizers = append(summarizers, s)
+	}
+
+	if s := newPolicyEvaluationSummarizer(b, ts); s != nil {
+		summarizers = append(summarizers, s)
+	}
+
+	return ctx.Poll(taskStageBackoffMin, taskStageBackoffMax, func(i int) (bool, error) {
+		options := tfe.TaskStageReadOptions{
+			Include: []tfe.TaskStageIncludeOpt{tfe.TaskStageTaskResults, tfe.PolicyEvaluationsTaskResults},
+		}
+		stage, err := b.client.TaskStages.Read(ctx.StopContext, stageID, &options)
+		if err != nil {
+			return false, generalError("Failed to retrieve task stage", err)
+		}
+
+		switch stage.Status {
+		case tfe.TaskStagePending:
+			// Waiting for it to start
+			return true, nil
+		// Note: Terminal statuses need to print out one last time just in case
+		case tfe.TaskStageRunning, tfe.TaskStagePassed:
+			for _, s := range summarizers {
+				cont, msg, err := s.Summarize(ctx, output, stage)
+				if err != nil {
+					errs.Append(err)
+					break
+				}
+				if !cont {
+					continue
+				}
+
+				// cont is true and we must continue to poll
+				if msg != nil {
+					output.OutputElapsed(*msg, len(*msg)) // Up to 2 digits are allowed by the max message allocation
+				}
+				return true, nil
+			}
+		case tfe.TaskStageCanceled, tfe.TaskStageErrored, tfe.TaskStageFailed:
+			for _, s := range summarizers {
+				cont, msg, err := s.Summarize(ctx, output, stage)
+				if err != nil {
+					errs.Append(err)
+					break
+				}
+
+				if !cont {
+					continue
+				}
+
+				// cont is true and we must continue to poll
+				if msg != nil {
+					output.OutputElapsed(*msg, len(*msg)) // Up to 2 digits are allowed by the max message allocation
+				}
+				return true, nil
+			}
+			return false, fmt.Errorf("Task Stage %s.", stage.Status)
+		case tfe.TaskStageAwaitingOverride:
+			for _, s := range summarizers {
+				cont, msg, err := s.Summarize(ctx, output, stage)
+				if err != nil {
+					errs.Append(err)
+					break
+				}
+				if !cont {
+					continue
+				}
+				// cont is true and we must continue to poll
+				if msg != nil {
+					output.OutputElapsed(*msg, len(*msg)) // Up to 2 digits are allowed by the max message allocation
+				}
+				return true, nil
+			}
```

Comment:
💅 There is a lot of duplication here which makes this method harder to follow than it needs to be. Specifically this part, duplicated 3 times. I wonder if you could encapsulate this in a new type that could be a little more explicit about the overall control flow here. It's hard to follow because the inner loop can print a status message or the summarizer can but it's unclear where exactly the responsibility lies.

```
for _, s := range summarizers {
	cont, msg, err := s.Summarize(ctx, output, stage)
	if err != nil {
		errs.Append(err)
		break
	}
	if !cont {
		continue
	}

	// cont is true and we must continue to poll
	if msg != nil {
		output.OutputElapsed(*msg, len(*msg)) // Up to 2 digits are allowed by the max message allocation
	}
	return true, nil
}
```